rascal-lsp/src/main/rascal/util/LanguageServer.rsc

```diff
@@ -82,45 +98,91 @@ alias CommandExecutor  = value (Command /*command*/);
+}
 alias Referrer         = set[loc] (loc /*origin*/, Tree /*fullTree*/, Tree /*lexicalAtCursor*/);
 
 @synopsis{Function profile for implementer contributions to a language server}
+@description{
+An implementer is called on-demand, when an implementation is requested by the IDE user.
+}
+@benefits{
+* is focused on a single implementation request, so does not need full program analysis.
+}
+@pitfalls{
+* should be extremely fast in order to provide interactive access.
+* careful use of `@memo` may help to cache dependencies, but this is tricky!
+}
 alias Implementer      = set[loc] (loc /*origin*/, Tree /*fullTree*/, Tree /*lexicalAtCursor*/);
 
 @synopsis{Each kind of service contibutes the implementation of one (or several) IDE features.}
 @description{
 Each LanguageService provides one aspect of definining the language server protocol.
-* a `parser` maps source code to a parse tree and indexes each part based on offset and length
-* a `summarizer` indexes a file as a ((Summary)), offering precomputed relations for looking up
-documentation, definitions, references, implementations and compiler errors and warnings.
-* a `outliner` maps a source file to a pretty hierarchy for visualization in the "outline" view
-* a `lenses` discovers places to add "lenses" (little views embedded in the editor) and connects commands to execute to each lense
-* an `inlayHinter` is like lenses but inbetween words
-* a `executor` executes the commands registered by `lenses` and `inlayHinters`
-* a `documenter` is a fast and location specific version of the `documentation` relation in a ((Summary)).
-* a `definer` is a fast and location specific version of the `definitions` relation in a ((Summary)).
-* a `referrer` is a fast and location specific version of the `references` relation in a ((Summary)).
-* an `implementer` is a fast and location specific version of the `implementations` relation in a ((Summary)).
+* ((parser)) maps source code to a parse tree and indexes each part based on offset and length
+* ((analyzer)) indexes a file as a ((Summary)), offering precomputed relations for looking up
+documentation, definitions, references, implementations and compiler errors and warnings. 
+   * ((analyzer))s focus on their own file, but may reuse cached or stored indices from other files.
+   * ((analyzer))s have to be quick since they run in an interactive editor setting.
+   * ((analyzer))s may store previous results (in memory) for incremental updates.
+   * ((analyzer))s are triggered during typing, in a short typing pause.
+* ((builder)) is similar to an `analyzer`, but it may perform computation-heavier additional checks.
+   * ((builder))s typically run whole-program analyses and compilation steps.
+   * ((builder))s have side-effects, they store generated code or code indices for future usage by the next build step, or by the next analysis step.
+   * ((builder))s are triggered on _save-file_ events; they _push_ information to an internal cache.
```

Comment:
add a warning: on-disk file changes outside of vs code do not trigger a builder (but will trigger an analyzer)